### PR TITLE
Python: deprecate 3.8

### PR DESCRIPTION
**Pull request by adamjstewart**
It's that time of year again! Python 3.8 has officially reached [EOL status](https://devguide.python.org/versions/). Since it is no longer supported by PSF and no longer receives fixes to security vulnerabilities, it should be deprecated.

Note that I am no longer maintaining the Python package and no longer removing these deprecated versions, I'm just marking them as deprecated to prevent people from accidentally installing them. This also helps to deprecate Python packages that only support Python 3.8 and older.

### CODE REVIEW

**Comment by adamjstewart** [comment]
@spackbot maintainers

**Comment by adamjstewart** [comment]
Could use some help getting visit to build with modern versions of Python.

**Comment by adamjstewart** [comment]
@spackbot maintainers

**Review by vicentebolea — APPROVED** [review]
adios2 looks good, we will fix the error that requires the conflicts after this PR

**Comment by adamjstewart** [comment]
@cyrush still looking for help adding newer versions of VisIt (supporting newer VTK and Python). Alternatively, we could remove VisIt from the E4S pipelines until it is supported.

**Comment by kwryankrattiger** [comment]
There is an updated version of VisIt available (3.4). It just needs to be added to Spack.

**Comment by adamjstewart** [comment]
I tried but couldn't get it to build. I'm not super familiar with CMake compared to Python, so maybe someone can take a look at the failing build in CI and provide hints to fix the build? Happy to try things out.

**Comment by cyrush** [comment]
The error looks like some confusion with looking for VTK-8 details -- even after VTK-9 was found.

Not sure why that is happening b/c the VTK version is being passed.

File: var/spack/repos/builtin/packages/visit/package.py
Comment (kwryankrattiger) [inline var/spack/repos/builtin/packages/visit/package.py]: I am digging into this in more detail here:

https://github.com/spack/spack/pull/47161

File: var/spack/repos/builtin/packages/bazel/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/bazel/package.py:57]: Can you use `with default_args(deprecated=True)` to reduce diff size by two thirds?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/bazel/package.py:57]: I'm going to delete all of these lines in less than a month, I don't want to bikeshed about their formatting.

Comment (haampie) [reply var/spack/repos/builtin/packages/bazel/package.py:57]: You are, but it stays in releases/v0.23 forever.

**Comment by adamjstewart** [comment]
@tgamblin would love to sneak this into 0.23. I think VisIt is the only thing holding up this PR. While we're waiting on #47161, we could either remove VisIt from CI or remove the Python 3.8 deprecation and only keep the secondary package deprecations (all of the things that will become deprecated once Python 3.8 is deprecated).

**Comment by adamjstewart** [comment]
This now passes all tests and has been open for over a month, giving everyone time to chime in if this affects them. Anything else needed for this to be merged?

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py:86]: Due to dependencies - the same versions in py-petsc4py, slepc, and py-slepc4py should also be deprecated.

cc: @joseeroman

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/petsc/package.py:86]: So far I've added explicit deprecations to:

1. Python 3.8
2. Direct dependencies on Python 3.8

I could keep adding 2nd and 3rd order deprecations, but this is a bit time consuming. The important thing is that installing any of these deprecated versions (including Python 3.8 itself) will raise a warning telling users that this version is deprecated and they should update. Not sure how far we want to keep going with this.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:86]: Ok - I can create a subsequent PR for these pkgs.

> The important thing is that installing any of these deprecated versions (including Python 3.8 itself) will raise a warning telling users that this version is deprecated and they should update

```console
balay@pj01:~/spack$ ./bin/spack spec slepc@3.13.0
==> Error: failed to concretize `slepc@3.13.0` for the following reasons:
     1. Cannot satisfy 'petsc@3.13'
     2. Cannot satisfy 'petsc@3.13'
        required because slepc depends on petsc@3.13 when @3.13 
          required because slepc@3.13.0 requested explicitly 
balay@pj01:~/spack$ ./bin/spack spec --deprecated slepc@3.13.0
==> Warning: using "petsc@3.13.6" which is a deprecated version
==> Warning: using "python@3.8.19" which is a deprecated version
==> Warning: using "cmake@3.28.2" which is a deprecated version
 -   slepc@3.13.0%gcc@14.2.1+arpack~blopex~cuda~hpddm~rocm build_system=generic arch=linux-fedora40-skylake
 -       ^arpack-ng@3.9.0%gcc@14.2.1~icb~ipo+mpi+shared build_system=cmake build_type=Release generator=make arch=linux-fedora40-skylake
```
Hm - spack is treating "deprecated" version as "removed" by default - so it won't be obvious to most users that the trigger here is due to a deprecated pkg in the dependency chain. [as they won't get the "deprecated version" warning]

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/petsc/package.py:86]: > I can create a subsequent PR for these pkgs.

You can also push to this PR if you want.

> it won't be obvious to most users that the trigger here is due to a deprecated pkg in the dependency chain

I don't like this, @alalazo @becker33 can we make this message more obvious (in a separate PR)?

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:86]: > You can also push to this PR if you want.

pushed. Thanks!

**Comment by balay** [comment]
https://gitlab.spack.io/spack/spack/-/jobs/13875522

```
==> Installing gdk-pixbuf-2.42.12-mbawvqqbntmenbd26tti3dzbjxpqj7ie [1/1]
==> Cannot find version 2.42.12 in url_list
==> Error: FetchError: All fetchers failed for spack-stage-gdk-pixbuf-2.42.12-mbawvqqbntmenbd26tti3dzbjxpqj7ie
        https://mirror.spack.io/_source-cache/archive/b9/b9505b3445b9a7e48ced34760c3bcb73e966df3ac94c95a148cb669ab748e3c7.tar.xz: DetailedHTTPError: GET https://mirror.spack.io/_source-cache/archive/b9/b9505b3445b9a7e48ced34760c3bcb73e966df3ac94c95a148cb669ab748e3c7.tar.xz returned 404: Not Found
    https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.42/gdk-pixbuf-2.42.12.tar.xz: DetailedHTTPError: GET https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.42/gdk-pixbuf-2.42.12.tar.xz returned 403: Forbidden
    https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.42/gdk-pixbuf-2.42.12.tar.xz: DetailedHTTPError: GET https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.40/gdk-pixbuf-2.42.12.tar.xz returned 403: Forbidden
```

This CI failure is unrelated to my changes -  cc: @eugeneswalker

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by eugeneswalker** [comment]
> https://gitlab.spack.io/spack/spack/-/jobs/13875522
> 
> ```
> ==> Installing gdk-pixbuf-2.42.12-mbawvqqbntmenbd26tti3dzbjxpqj7ie [1/1]
> ==> Cannot find version 2.42.12 in url_list
> ==> Error: FetchError: All fetchers failed for spack-stage-gdk-pixbuf-2.42.12-mbawvqqbntmenbd26tti3dzbjxpqj7ie
>         https://mirror.spack.io/_source-cache/archive/b9/b9505b3445b9a7e48ced34760c3bcb73e966df3ac94c95a148cb669ab748e3c7.tar.xz: DetailedHTTPError: GET https://mirror.spack.io/_source-cache/archive/b9/b9505b3445b9a7e48ced34760c3bcb73e966df3ac94c95a148cb669ab748e3c7.tar.xz returned 404: Not Found
>     https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.42/gdk-pixbuf-2.42.12.tar.xz: DetailedHTTPError: GET https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.42/gdk-pixbuf-2.42.12.tar.xz returned 403: Forbidden
>     https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.42/gdk-pixbuf-2.42.12.tar.xz: DetailedHTTPError: GET https://ftp.acc.umu.se/pub/gnome/sources/gdk-pixbuf/2.40/gdk-pixbuf-2.42.12.tar.xz returned 403: Forbidden
> ```
> 
> This CI failure is unrelated to my changes - cc: @eugeneswalker

What should we do about this? It looks like source can't be fetched, unclear what to do here.

**Comment by adamjstewart** [comment]
I'm seeing that error on multiple PRs, I think CI is broken.

**Comment by kwryankrattiger** [comment]
Fixing here: https://github.com/spack/spack/pull/47823

The mirror location seems to have been removed. Replacing with gitlab unless there is a better option someone knows about.

**Comment by balay** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
Let's merge #47719 first, which currently waits for macOS runners to be available again in Gitlab.

**Comment by adamjstewart** [comment]
@haampie can this be merged now?

**Review by eugeneswalker — APPROVED** [review]

**Pull request merged**
